include/soci/soci-platform.h

```diff
@@ -97,7 +97,7 @@ namespace std {
 // mode, we just need to check for the minimal compiler version supporting them
 // (see https://msdn.microsoft.com/en-us/library/hh567368.aspx).
 
-#if defined(SOCI_HAVE_CXX_C11) || (defined(_MSC_VER) && _MSC_VER >= 1800)
+#if defined(SOCI_HAVE_CXX_C11) || (defined(_MSC_VER) && _MSC_VER >= 1800) || (__cplusplus >= 201103L)
```

Comment:
Sorry, I don't understand the changes in this file. `SOCI_HAVE_CXX_C11` should be already defined if `__cplusplus >= 201103L` in `include/soci/soci-config.h`. It doesn't make sense to test for both conditions here, as they're supposed to be equivalent (with the exception of the special case of MSVC without `/std:latest` or similar).

Reply:
> Sorry, I don't understand the changes in this file. SOCI_HAVE_CXX_C11 should be already defined if __cplusplus >= 201103L in include/soci/soci-config.h

how? by user input? Its not for me. Tell me through which mechanism

Reply:
> > Sorry, I don't understand the changes in this file. SOCI_HAVE_CXX_C11 should be already defined if __cplusplus >= 201103L in include/soci/soci-config.h
> 
> how? by user input? Its not for me. Tell me through which mechanism.

Sorry, I forgot that I modified `soci/soci-config.h.in` locally to define `SOCI_HAVE_CXX_C11` if `__cplusplus >= 201103L`, but this is not in the upstream sources. IMO it could be, but I don't know what @mloskot thinks about it.

> Its not if I run cmake as:
> 
> `cmake .. -DCMAKE_CXX_STANDARD=17`

This should define `SOCI_HAVE_CXX_C11` somehow. If we don't want to add
```cpp
#if __cplusplus >= 201103L
    #define SOCI_HAVE_CXX_C11
#endif
```
to `soci-config.h.in`, then CMake should add `SOCI_HAVE_CXX_C11` to `CONFIGURED_VARIABLES` in this case.

Reply:
> Sorry, I forgot that I modified `soci/soci-config.h.in` locally to define `SOCI_HAVE_CXX_C11`
> if `__cplusplus >= 201103L`, but this is not in the upstream sources.
> IMO it could be, but I don't know what @mloskot thinks about it.

I think it would be useful to define it.

Also, in https://github.com/SOCI/soci/issues/752#issue-496199995 I outlined some obvious rules about relation between `CMAKE_CXX_STANDARD` and `SOCI_CXX_C11`. I forgot about `SOCI_HAVE_CXX_C11 `, which seems to fall into the same category.

Reply:
I don't know anything about CMake (except that it's very confusing), but I don't think `SOCI_HAVE_CXX_C11` is really in the same category: it's supposed to be defined if C++11 is available (this is the usual convention for `HAVE` symbols), i.e. it's not directly under the user control, while `SOCI_CXX_C11` is a user-settable option.

OTOH I do see
```cmake
set(SOCI_HAVE_CXX_C11 ${SOCI_CXX_C11} CACHE INTERNAL "Enables C++11 support")
```
in `cmake/SociConfig.cmake`, so it looks like `SOCI_HAVE_CXX_C11` is already set from `SOCI_CXX_C11`, but for some reason it doesn't end up in `soci-config.h`. Any idea why?

Reply:
All three, `CMAKE_CXX_STANDARD` and `SOCI_CXX_C11 and `SOCI_HAVE_CXX_C11`, related to selecting C++11 and that is what I meant. Source of setting of those, user or inference, does not matter really. Shortly, they all three need to be taken into account when any logic is being changed that is related to setting C++11.

>(...) so it looks like SOCI_HAVE_CXX_C11 is already set from SOCI_CXX_C11, but for some reason it doesn't end up in soci-config.h. Any idea why?

Must be because `SOCI_CXX_C11=OFF`.
If you request C++11 like so then `SOCI_HAVE_CXX_C11` is defined:

```
cmake -DSOCI_CXX_C11=ON ..
grep 11 include/soci/soci-config.h
// Enables C++11 support
#define SOCI_HAVE_CXX_C11
```

Reply:
But the OP says that it's not set after `cmake .. -DCMAKE_CXX_STANDARD=17`, so do we just need to set `SOCI_CXX_C11=ON` if `CMAKE_CXX_STANDARD` is set to 17 (or 14, 11, 20, ...)?

As I said, I don't really know what is the CMake way of handling this. But I still stand by my original comment, i.e. that it doesn't make sense to test both `__cplusplus` and `SOCI_HAVE_CXX_C11` as they're supposed to be the same thing.

BTW, why do we have the 2nd `C` in `CXX_C11`? It doesn't seem to stand for anything.

Reply:
> But the OP says that it's not set after cmake .. -DCMAKE_CXX_STANDARD=17,

Yes, it is not set because such setting is not implemented.
As far as I understand, it is being / or will be proposed to implement such enhancement to make the three variables set together as necessary. That is why I documented the need for such enhancement as #752 referring to @5cript 's initial comments and idea.

>  so do we just need to set SOCI_CXX_C11=ON if CMAKE_CXX_STANDARD is set to 17 (or 14, 11, 20, ...)?

Yes, I think so. Again, that is what I described in RFC-like issue #752 which I opened as a guide and discussion thread about implementation of improved handling of `SOCI_CXX_C11`, `CMAKE_CXX_STANDARD ` and related variables.

> BTW, why do we have the 2nd C in CXX_C11? It doesn't seem to stand for anything.

AFAIS, it was proposed like that in #327 and I did not object the name (or I missed the `C`).
I don't mind renaming, for example, to `SOCI_CXX_11`.

Reply:
I am not qualified to talk about anything CMake-related, so I won't, but IMO it would indeed be better to rename all symbols with `CXX_C11` in their names to use just `CXX11` (or `CXX_11` if you prefer).

Reply:
@vadz I have renamed `SOCI_CXX_C11` to `SOCI_CXX11` in #757

Next step will be to add support for `CMAKE_CXX_STANDARD=11` and later values. I think I can do it as well.

Reply:
FYI, `CMAKE_CXX_STANDARD` support has now been added. So, this PR could be re-worked and updated to add support for `__cplusplus`.  If it does not happen within next 1-2 weeks, it will be the idea may miss 4.0.0 release.